### PR TITLE
Consistent and intuitive description for returned data cubes #149

### PR DESCRIPTION
**Pull request by m-mohr**
See issue #149. Tries to fix and improve the data cube descriptions for the returned cubes.
Also removed some other not so useful texts.

### CODE REVIEW

File: add_dimension.json [hunk 1]
```diff
@@ -50,7 +50,7 @@
         }
     ],
     "returns": {
-        "description": "The data cube with a newly added dimension.",
+        "description": "The data cube with a newly added dimension. The new dimension has exactly one dimension label. All other dimensions have not changed.",
```
Comment (soxofaan) [inline add_dimension.json:53]: > All other dimensions have not changed.

Maybe it's better to check with a native English speaker, but the past tense here in a `returns` doc sounds bit strange to me. Maybe it's easier to just say "are unchanged" or "do not change"

File: load_collection.json [hunk 2]
```diff
@@ -208,7 +208,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube for further processing.",
+        "description": "A data cube for further processing. The dimensions and dimension properties (name, type, labels, reference system and resolution) correspond to the collection's metadata, but the dimension labels got restricted as specified in the parameters.",
```
Comment (soxofaan) [inline load_collection.json:211]: I think  `labels got restricted` can be just `labels are restricted`

File: ndvi.json [hunk 3]
```diff
@@ -52,7 +52,7 @@
         }
     ],
     "returns": {
-        "description": "A raster data cube containing the computed NDVI values.",
+        "description": "A raster data cube containing the computed NDVI values. Depending on the value passed to `target_band` the data cube has changed differently:\n\n* `target_band` is `null`: The data cube is missing the dimension of type `bands`, the number of dimensions decreased by one. The dimension properties (name, type, labels, reference system and resolution) for all other dimensions have not changed.\n* `target_band` is a string: The data cube has the same dimensions. The dimension properties have not changed, but  the number of dimension labels for the dimension of type `bands` has increased by one. The additional label has the name specified in `target_band`.",
```
Comment (soxofaan) [inline ndvi.json:55]: as noted above (and depending on what is decided): I think the tense of the verbs can be tweaked a bit here: now it's a bit a mix: `is missing`, `decreased`, `have not changed`, `has`, `has increased`

Comment (m-mohr) [reply ndvi.json:55]: Thanks, indeed I'm not 100% sure about the tenses. Could also be future "a dimension will be missing". Native speakers here?

File: add_dimension.json (hunk 1 above)
Comment (przell) [reply add_dimension.json:53]: > All other dimensions have not changed.

> "are unchanged" or "do not change"

What about:
All other dimensions remain unchanged.

File: load_collection.json (hunk 2 above)
Comment (przell) [reply load_collection.json:211]: `lables are restricted` sounds good to me.

File: add_dimension.json (hunk 1 above)
Comment (m-mohr) [reply add_dimension.json:53]: That would work in this case, but there are more phrases used in the past tense, so we need to clarify the tense we want to use anyway.

File: ndvi.json (hunk 3 above)
Comment (przell) [reply ndvi.json:55]: I'd stick to present tense. 
My go would be:

"description": "A raster data cube containing the computed NDVI values. 
The data cube results differently depending on the value passed to `target_band`:
\n\n* `target_band` is `null`: The data cube does not contain the dimension of type `bands` anymore, the number of dimensions decreases by one. The dimension properties (name, type, labels, reference system and resolution) for all other dimensions remain unchanged.
\n* `target_band` is a string: The data cube keeps the same dimensions. The dimension properties remain unchanged, but the number of dimension labels for the dimension of type `bands` increases by one. The additional label is named as specified in `target_band`."

File: aggregate_temporal.json [hunk 4]
```diff
@@ -136,7 +136,7 @@
         },
         {
             "name": "dimension",
-            "description": "The name of the temporal dimension for aggregation. All data along the dimension will be passed through the specified reducer. If the dimension is not set or set to `null`, the data cube is expected to only have one temporal dimension. Fails with a `TooManyDimensions` error if it has more dimensions. Fails with a `DimensionNotAvailable` error if the specified dimension does not exist.\n\n**Note:** The default dimensions a data cube provides are described in the collection's metadata field `cube:dimensions`.",
+            "description": "The name of the temporal dimension for aggregation. All data along the dimension will be passed through the specified reducer. If the dimension is not set or set to `null`, the data cube is expected to only have one temporal dimension. Fails with a `TooManyDimensions` error if it has more dimensions. Fails with a `DimensionNotAvailable` error if the specified dimension does not exist.",
```
Comment (przell) [inline aggregate_temporal.json:139]: > All data along the dimension will be passed through the specified reducer.

All data along the dimension will is through the specified reducer.

File: aggregate_temporal.json [hunk 5]
```diff
@@ -157,7 +157,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with potentially lower resolution and potentially lower cardinality, but the same number of dimensions as the original data cube.",
+        "description": "A data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed, except for the resolution and dimension labels of the given temporal dimension.",
```
Comment (przell) [inline aggregate_temporal.json:160]: > have not changed

remain unchanged

File: aggregate_temporal.json (hunk 4 above)
Comment (przell) [inline aggregate_temporal.json:139]: > All data along the dimension will be passed through the specified reducer.

All data along the dimension is passed through the specified reducer.

File: aggregate_temporal.json (hunk 5 above)
Comment (przell) [inline aggregate_temporal.json:160]: > have not changed

remain unchanged

File: aggregate_temporal_period.json [hunk 6]
```diff
@@ -67,7 +67,7 @@
         },
         {
             "name": "dimension",
-            "description": "The name of the temporal dimension for aggregation. All data along the dimension will be passed through the specified reducer. If the dimension is not set or set to `null`, the data cube is expected to only have one temporal dimension. Fails with a `TooManyDimensions` error if it has more dimensions. Fails with a `DimensionNotAvailable` error if the specified dimension does not exist.\n\n**Note:** The default dimensions a data cube provides are described in the collection's metadata field `cube:dimensions`.",
+            "description": "The name of the temporal dimension for aggregation. All data along the dimension will be passed through the specified reducer. If the dimension is not set or set to `null`, the data cube is expected to only have one temporal dimension. Fails with a `TooManyDimensions` error if it has more dimensions. Fails with a `DimensionNotAvailable` error if the specified dimension does not exist.",
```
Comment (przell) [inline aggregate_temporal_period.json:70]: > All data along the dimension will be passed through the specified reducer.

All data along the dimension is passed through the specified reducer.

File: aggregate_temporal_period.json [hunk 7]
```diff
@@ -88,7 +88,7 @@
...
+        "description": "A data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed, except for the resolution and dimension labels of the given temporal dimension. The specified temporal dimension has the following dimension labels (`YYYY` = four-digit year, `MM` = two-digit month, `DD` two-digit day of month):\n\n* `hour`: `YYYY-MM-DD-00` - `YYYY-MM-DD-23`\n* `day`: `YYYY-001` - `YYYY-365`\n* `week`: `YYYY-01` - `YYYY-52`\n* `dekad`: `YYYY-00` - `YYYY-36`\n* `month`: `YYYY-01` - `YYYY-12`\n* `season`: `YYYY-djf` (December - February), `YYYY-mam` (March - May), `YYYY-jja` (June - August), `YYYY-son` (September - November).\n* `tropical-season`: `YYYY-ndjfma` (November - April), `YYYY-mjjaso` (May - October).\n* `year`: `YYYY`\n* `decade`: `YYY0`\n* `decade-ad`: `YYY1`",
```
Comment (przell) [inline aggregate_temporal_period.json:91]: > have not changed

remain unchanged

File: anomaly.json [hunk 8]
```diff
@@ -46,7 +46,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube. The cardinality, resolution, the number of dimensions and the dimension labels are the same as for the original data cube.",
+        "description": "A data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed.",
```
Comment (przell) [inline anomaly.json:49]: > have not changed.

remain unchanged

File: apply.json [hunk 9]
```diff
@@ -51,7 +51,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with the newly computed values. The resolution, cardinality and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube with the newly computed values and the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed.",
```
Comment (przell) [inline apply.json:54]: > have not changed.

remain unchanged

File: apply_kernel.json [hunk 10]
```diff
@@ -70,7 +70,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with the newly computed values. The resolution, cardinality and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube with the newly computed values and the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed.",
```
Comment (przell) [inline apply_kernel.json:73]: remain unchanged

File: climatological_normal.json [hunk 11]
```diff
@@ -52,7 +52,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with potentially cardinality, but the same resolution and number of dimensions as the original data cube. The temporal dimension has the following dimension labels:\n\n* `day`: `001` - `365`\n* `month`: `01` - `12`\n* `climatology-period`: `climatology-period`\n* `season`: `djf` (December - February), `mam` (March - May), `jja` (June - August), `son` (September - November)\n* `tropical-season`: `ndjfma` (November - April), `mjjaso` (May - October)",
+        "description": "A data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed, except for the resolution and dimension labels of the temporal dimension. The temporal dimension has the following dimension labels:\n\n* `day`: `001` - `365`\n* `month`: `01` - `12`\n* `climatology-period`: `climatology-period`\n* `season`: `djf` (December - February), `mam` (March - May), `jja` (June - August), `son` (September - November)\n* `tropical-season`: `ndjfma` (November - April), `mjjaso` (May - October)",
```
Comment (przell) [inline climatological_normal.json:55]: > have not changed

remain unchanged

File: drop_dimension.json [hunk 12]
```diff
@@ -23,7 +23,7 @@
         }
     ],
     "returns": {
-        "description": "The data cube with one dimension less.",
+        "description": "A data cube without the specified dimension. The number of dimensions decreased by one, but the dimension properties (name, type, labels, reference system and resolution) for all other dimensions have not changed.",
```
Comment (przell) [inline drop_dimension.json:26]: > The number of dimensions decreased by one

The number of dimensions decreases by one

> for all other dimensions have not changed.

remain unchanged

File: filter_bbox.json [hunk 13]
```diff
@@ -91,7 +91,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube restricted to the bounding box. Therefore, the cardinality is potentially lower, but the resolution and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube restricted to the bounding box. The dimensions and dimension properties (name, type, labels, reference system and resolution) have not changed, except that the spatial dimensions have less (or the same) dimension labels.",
```
Comment (przell) [inline filter_bbox.json:94]: > have not changed

remain unchanged

File: filter_labels.json [hunk 14]
```diff
@@ -64,7 +64,7 @@
         }
     ],
     "returns": {
-        "description": "The filtered data cube. Therefore, the cardinality of the dimension is potentially lower, but the resolution and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed, except that the given dimension has less (or the same) dimension labels.",
```
Comment (przell) [inline filter_labels.json:67]: > have not changed

remain unchanged

File: filter_spatial.json [hunk 15]
```diff
@@ -25,7 +25,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube restricted to the specified geometries. Therefore, the cardinality is potentially lower, but the resolution and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube restricted to the specified geometries. The dimensions and dimension properties (name, type, labels, reference system and resolution) have not changed, except that the spatial dimensions have less (or the same) dimension labels.",
```
Comment (przell) [inline filter_spatial.json:28]: > have not changed

remain unchanged

File: filter_temporal.json [hunk 16]
```diff
@@ -73,7 +73,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube restricted to the specified temporal extent. Therefore, the cardinality is potentially lower, but the resolution and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube restricted to the specified temporal extent. The dimensions and dimension properties (name, type, labels, reference system and resolution) have not changed, except that the given temporal dimension(s) have less (or the same) dimension labels.",
```
Comment (przell) [inline filter_temporal.json:76]: > have not changed

remain unchanged

File: mask.json [hunk 17]
```diff
@@ -38,7 +38,7 @@
         }
     ],
     "returns": {
-        "description": "The masked raster data cube.",
+        "description": "A masked raster data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed.",
```
Comment (przell) [inline mask.json:41]: > have not changed

remain unchanged

File: reduce_dimension.json [hunk 18]
```diff
@@ -63,7 +63,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with the newly computed values. It has one dimension less than the original data cube, but the remaining dimensions have their resolution and cardinality preserved.",
+        "description": "A data cube with the newly computed values. It is missing the given dimension, the number of dimensions decreased by one. The dimension properties (name, type, labels, reference system and resolution) for all other dimensions have not changed.",
```
Comment (przell) [inline reduce_dimension.json:66]: > have not changed

remain unchanged

File: reduce_dimension_binary.json [hunk 19]
```diff
@@ -67,7 +67,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with the newly computed values. It has one dimension less than the original data cube, but the remaining dimensions have their resolution and cardinality preserved.",
+        "description": "A data cube with the newly computed values. It is missing the given dimension, the number of dimensions decreased by one. The dimension properties (name, type, labels, reference system and resolution) for all other dimensions have not changed.",
```
Comment (przell) [inline reduce_dimension_binary.json:70]: > It is missing the given dimension, the number of dimensions decreased by one.

It does not contain the given dimension anymore, the number of dimension decreases by one.

> have not changed

remain unchanged

File: rename_labels.json [hunk 20]
```diff
@@ -59,7 +59,7 @@
         }
     ],
     "returns": {
-        "description": "The data cube with the renamed labels. The old labels can not be referenced any longer.",
+        "description": "The data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed, except that for the given dimension the labels have been changed. The old labels can not be referred to any longer. The number of labels has not changed.",
```
Comment (przell) [inline rename_labels.json:62]: > have not changed, except that for the given dimension the labels have been changed

remain the same, except that for the given dimension the labels change

> The number of labels has not changed.

The number of labels remains the same

File: resample_cube_spatial.json [hunk 21]
```diff
@@ -48,7 +48,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with potentially lower spatial resolution and potentially lower cardinality, but the same number of dimensions as the original data cube.",
+        "description": "A data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed, except for the resolution and dimension labels of the spatial dimensions.",
```
Comment (przell) [inline resample_cube_spatial.json:51]: > have not changed

remain unchanged

File: resample_cube_temporal.json [hunk 22]
```diff
@@ -76,7 +76,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube with potentially lower temporal resolution and potentially lower cardinality, but the same number of dimensions as the original data cube.",
+        "description": "A raster data cube with the same dimensions and the same dimension properties (name, type, labels, reference system and resolution) for all non-temporal dimensions. For the temporal dimension the name and type have not changed, but the reference system has changed and the labels and resolution may have changed.",
```
Comment (przell) [inline resample_cube_temporal.json:79]: > dimension the name and type have not changed, but the reference system has changed and the labels and resolution may have changed.

dimension the name and type remain unchanged, but the reference system changes and the labels and resolution may change.

File: run_udf.json [hunk 23]
```diff
@@ -93,7 +93,7 @@
         }
     },
     "returns": {
-        "description": "The data processed by the UDF. Returns a raster data cube if a raster data cube was passed for `data`. If an array was passed for `data`, the returned value is defined by the context and is exactly what the UDF returned.",
+        "description": "The data processed by the UDF.\n\n* Returns a raster data cube, if a raster data cube was passed for `data`. Details on the dimensions and dimension properties (name, type, labels, reference system and resolution) depend on the UDF.\n* If an array was passed for `data`, the returned value can be of any data type, but is exactly what the UDF returned.",
```
Comment (przell) [inline run_udf.json:96]: >  was passed for `data`

is passed for `data`

> If an array was passed for `data`, the returned value can be of any data type, but is exactly what the UDF returned.

If an array is passed for `data`, the returned value can be of any data type, but is exactly what the UDF returns.

File: run_udf_externally.json [hunk 24]
```diff
@@ -52,7 +52,7 @@
         }
     ],
     "returns": {
-        "description": "The data processed by the UDF service. Returns a raster data cube if a raster data cube was passed for `data`. If an array was passed for `data`, the returned value is defined by the context and is exactly what the UDF returned.",
+        "description": "The data processed by the UDF service.\n\n* Returns a raster data cube, if a raster data cube was passed for `data`. Details on the dimensions and dimension properties (name, type, labels, reference system and resolution) depend on the UDF.\n* If an array was passed for `data`, the returned value can be of any data type, but is exactly what the UDF returned.",
```
Comment (przell) [inline run_udf_externally.json:55]: > if a raster data cube was passed for `data`

if a raster data cube is passed for `data`

> If an array was passed for `data`, the returned value can be of any data type, but is exactly what the UDF returned.

If an array is passed for `data`, the returned value can be of any data type, but is exactly what the UDF returns.

File: trim_cube.json [hunk 25]
```diff
@@ -16,7 +16,7 @@
         }
     ],
     "returns": {
-        "description": "A trimmed raster data cube.",
+        "description": "A trimmed raster data cube with the same dimensions. The dimension properties name, type, reference system and resolution have not changed. The number of dimension labels may have decreased.",
```
Comment (przell) [inline trim_cube.json:19]: > have not changed

remain unchanged

> may have decreased

may decrease

File: filter_bands.json [hunk 26]
```diff
@@ -59,7 +59,7 @@
         }
     ],
     "returns": {
-        "description": "A data cube limited to a subset of its original bands. Therefore, the cardinality is potentially lower, but the resolution and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube limited to a subset of its original bands. The dimensions and dimension properties (name, type, labels, reference system and resolution) have not changed, except that the dimension of type `bands` has less (or the same) dimension labels.",
```
Comment (przell) [inline filter_bands.json:62]: > have not changed

remain unchanged

File: mask_polygon.json [hunk 27]
```diff
@@ -48,7 +48,7 @@
         }
     ],
     "returns": {
-        "description": "The masked raster data cube.",
+        "description": "A masked raster data cube with the same dimensions. The dimension properties (name, type, labels, reference system and resolution) have not changed.",
```
Comment (przell) [inline mask_polygon.json:51]: > have not changed

remain unchanged

File: rename_dimension.json [hunk 28]
```diff
@@ -30,7 +30,7 @@
         }
     ],
     "returns": {
-        "description": "The data cube with the renamed dimension. The old name can not be referenced any longer.",
+        "description": "A data cube with the same dimensions, but the name of one of the dimensions has changed. The old name can not be referred to any longer. The dimension properties (name, type, labels, reference system and resolution) have not changed.",
```
Comment (przell) [inline rename_dimension.json:33]: > has changed

changes

> have not changed.

remain unchanged.

File: resample_spatial.json [hunk 29]
```diff
@@ -110,7 +110,7 @@
         }
     ],
     "returns": {
-        "description": "A raster data cube with values warped onto the new projection.",
+        "description": "A raster data cube with values warped onto the new projection. It has the same dimensions and the same dimension properties (name, type, labels, reference system and resolution) for all non-spatial or vertical spatial dimensions. For the horizontal spatial dimensions the name and type have not changed, but the reference system has changed and the labels and resolution may have changed.",
```
Comment (przell) [inline resample_spatial.json:113]: > For the horizontal spatial dimensions the name and type have not changed, but the reference system has changed and the labels and resolution may have changed.

For the horizontal spatial dimensions the name and type remain unchanged, but the reference system changes and the labels and resolution may change.

File: apply_dimension.json [hunk 30]
```diff
@@ -74,7 +74,7 @@
...
...is not** the target dimension and the latter **exists**:\n   * The number of dimensions decreased by one as the source dimension has been dropped.\n   * The target dimension properties name, type and reference system have not changed.\n   * The resolution has changed, the number of dimension labels is equal to the number of values computed by the process and the dimension labels are incrementing integers starting from zero\n3. The source dimension **is not** the target dimension and the latter **does not exist**:\n   * The number of dimensions have not changed, but the source dimension has been replaced with the target dimension.\n   * The target dimension has the specified name and the type `other`. The reference system has not been changed.\n   * The resolution has changed, the number of dimension labels is equal to the number of values computed by the process and the dimension labels are incrementing integers starting from zero\n\nFor all three cases except for the exception in the first case, the resolution has changed, the number of dimension labels is equal to the number of values computed by the process and the dimension labels are incrementing integers starting from zero.",
```
Comment (przell) [inline apply_dimension.json:77]: "description": "A data cube with the newly computed values. All dimensions stay the same, except for the dimensions specified in corresponding parameters. There are three cases how the data cube changes:
\n
\n1. The source dimension **is** the target dimension:
\n   * The (number of) dimensions remain unchanged.
\n   * The source dimension properties name, type and reference system remain unchanged.
\n   * The dimension labels and the resolution are preserved when the number of pixel values in the source dimension is equal to the number of values computed by the process. The other case is described below.

\n2. The source dimension **is not** the target dimension and the latter **exists**:
\n   * The number of dimensions decreases by one as the source dimension is dropped.
\n   * The target dimension properties name, type and reference system remain unchanged.
\n   * The resolution changes, the number of dimension labels is equal to the number of values computed by the process and the dimension labels are incrementing integers starting from zero

\n3. The source dimension **is not** the target dimension and the latter **does not exist**:
\n   * The number of dimensions do not changed, but the source dimension is replaced with the target dimension.
\n   * The target dimension has the specified name and the type `other`. The reference system is not changed.
\n   * The resolution changes, the number of dimension labels is equal to the number of values computed by the process and the dimension labels are incrementing integers starting from zero
\n
\nFor all three cases except for the exception in the first case, the resolution changes, the number of dimension labels is equal to the number of values computed by the process and the dimension labels are incrementing integers starting from zero.

**Comment by m-mohr** [comment]
@przell Oh, wow, danke! Werde ich mich morgen drum kümmern, wenn kein Veto aus dem Projekt kommt.

File: resample_spatial.json (hunk 29 above)
Comment (lforesta) [reply resample_spatial.json:113]: The reference system does not have to change if one only want to resample the data, which is allowed by the process. I propose:
"For the horizontal spatial dimensions the name and type remain unchanged, but reference system, labels and resolution may change."

Comment (m-mohr) [reply resample_spatial.json:113]: Good point, @lforesta.

Will extend it a bit to "For the horizontal spatial dimensions the name and type remain unchanged, but reference system, labels and resolution may change depending on the given parameters."

File: ndvi.json (hunk 3 above)
Comment (m-mohr) [reply ndvi.json:55]: @przell 

> The data cube results differently depending on the value passed to target_band:" sounds incorrect.

Should it be: 

> The data cube results differ depending on the value passed to target_band:

File: apply_dimension.json (hunk 30 above)
Comment (m-mohr) [reply apply_dimension.json:77]: @przell I guess "do not changed" should be "remain unchanged"

**Comment by m-mohr** [comment]
Applied the changes as suggested, with two or three changes as mentioned in the comments. Thanks for the suggestions!

**Pull request merged**